File: sdk/consensus/fork_choice.py
"""
Fork Choice Rule Implementation (GHOST Algorithm).

Matches luxtensor-consensus/src/fork_choice.rs.

Implements Greedy Heaviest-Observed Sub-Tree (GHOST) algorithm
for selecting the canonical chain head during forks.

Key concepts:
- GHOST prefers the subtree with most accumulated weight/work
- Score = accumulated stake-weighted attestations
- Head = block with highest score among tips
"""

from dataclasses import dataclass, field
from typing import Dict, List, Optional, Set
from collections import deque
import threading


@dataclass
class BlockInfo:
    """
    Minimal block information for fork choice.

    Note: This is a simplified representation for SDK use.
    Full block data should be retrieved from chain storage.
    """
    hash: str           # Block hash (0x prefixed)
    parent_hash: str    # Parent block hash
    height: int         # Block height/number
    timestamp: int      # Unix timestamp
    proposer: str       # Validator who proposed the block

    def __hash__(self):
        return hash(self.hash)

    def __eq__(self, other):
        if isinstance(other, BlockInfo):
            return self.hash == other.hash
        return False


class ForkChoiceError(Exception):
    """Errors from fork choice operations."""
    pass


class ForkChoice:
    """
    Fork choice rule implementation (GHOST algorithm).

    Matches luxtensor-consensus ForkChoice struct.

    GHOST (Greedy Heaviest-Observed Sub-Tree) selects the chain
    that has the most accumulated validator support.

    Usage:
        genesis = BlockInfo(
            hash="0x...",
            parent_hash="0x" + "00" * 32,
            height=0,
            timestamp=0,
            proposer="0x..."
        )
        fc = ForkChoice(genesis)

        # Add blocks as they arrive
        fc.add_block(new_block)

        # Get current head
        head = fc.get_head()
    """

    def __init__(self, genesis: BlockInfo):
        """Create a new fork choice instance with genesis block."""
        self._blocks: Dict[str, BlockInfo] = {}
        self._children: Dict[str, List[str]] = {}  # parent_hash -> [child_hashes]
        self._scores: Dict[str, int] = {}          # hash -> score
        self._head_hash: str = genesis.hash
        self._lock = threading.RLock()

        # Initialize with genesis
        self._blocks[genesis.hash] = genesis
        self._children[genesis.hash] = []
        self._scores[genesis.hash] = 1  # Genesis has base score of 1

    def add_block(self, block: BlockInfo) -> None:
        """
        Add a new block to the fork choice.

        Args:
            block: Block to add

        Raises:
            ForkChoiceError: If block is duplicate or orphan
        """
        with self._lock:
            # Check for duplicate
            if block.hash in self._blocks:
                raise ForkChoiceError(f"Block {block.hash[:16]}... already exists")

            # Check parent exists
            if block.parent_hash not in self._blocks:
                raise ForkChoiceError(
                    f"Orphan block: parent {block.parent_hash[:16]}... not found"
                )

            # Add block
            self._blocks[block.hash] = block
            self._children[block.hash] = []

            # Add to parent's children
            if block.parent_hash not in self._children:
                self._children[block.parent_hash] = []
            self._children[block.parent_hash].append(block.hash)

            # Calculate score (parent's score + 1)
            parent_score = self._scores.get(block.parent_hash, 0)
            self._scores[block.hash] = parent_score + 1

            # Update head if this block has higher score
            self._update_head()

    def get_head(self) -> BlockInfo:
        """Get the current head block."""
        with self._lock:
            if self._head_hash not in self._blocks:
                raise ForkChoiceError("Head block not found")
            return self._blocks[self._head_hash]

    def get_head_hash(self) -> str:
        """Get the head hash."""
        with self._lock:
            return self._head_hash

    def get_block(self, hash: str) -> Optional[BlockInfo]:
        """Get a block by hash."""
        with self._lock:
            return self._blocks.get(hash)

    def get_canonical_chain(self) -> List[BlockInfo]:
        """
        Get the canonical chain from genesis to head.

        Returns:
            List of blocks from genesis to head (inclusive)
        """
        with self._lock:
            chain = []
            current_hash = self._head_hash

            while current_hash in self._blocks:
                block = self._blocks[current_hash]
                chain.append(block)

                # Stop at genesis (parent not in blocks)
                if block.parent_hash not in self._blocks:
                    break
                current_hash = block.parent_hash

            chain.reverse()  # Genesis first
            return chain

    def get_score(self, hash: str) -> Optional[int]:
        """Get the score of a block."""
        with self._lock:
            return self._scores.get(hash)

    def has_block(self, hash: str) -> bool:
        """Check if a block exists."""
        with self._lock:
            return hash in self._blocks

    @property
    def block_count(self) -> int:
        """Get the number of blocks in the fork choice."""
        with self._lock:
            return len(self._blocks)

    def _update_head(self) -> None:
        """Update the head to the block with the highest score (GHOST algorithm)."""
        best_hash = self._head_hash
        best_score = self._scores.get(self._head_hash, 0)

        for hash, score in self._scores.items():
            # Prefer higher score, or same score with later block
            if score > best_score:
                best_score = score
                best_hash = hash
            elif score == best_score and hash > best_hash:
                # Tie-break by hash (deterministic)
                best_hash = hash

        self._head_hash = best_hash

    def recompute_scores(self) -> None:
        """
        Recompute scores for all blocks (useful after reorganization).

        Uses BFS from genesis for correct topological ordering.
        """
        with self._lock:
            # Find genesis (block with no parent in our set)
            genesis_hash = None
            for hash, block in self._blocks.items():
                if block.parent_hash not in self._blocks:
                    genesis_hash = hash
                    break

            if genesis_hash is None:
                return

            # BFS from genesis
            self._scores = {genesis_hash: 1}
            queue = deque([genesis_hash])

            while queue:
                current_hash = queue.popleft()
                current_score = self._scores[current_hash]

                for child_hash in self._children.get(current_hash, []):
                    self._scores[child_hash] = current_score + 1
                    queue.append(child_hash)

            self._update_head()

    def get_blocks_at_height(self, height: int) -> List[BlockInfo]:
        """Get all blocks at a specific height."""
        with self._lock:
            return [
                block for block in self._blocks.values()
                if block.height == height
            ]

    def prune(self, keep_depth: int) -> int:
        """
        Prune old blocks (keep only canonical chain + recent forks).

        Args:
            keep_depth: Number of blocks from head to keep

        Returns:
            Number of blocks pruned
        """
        with self._lock:
            head = self.get_head()
            min_height = max(0, head.height - keep_depth)

            # Build set of canonical hashes to keep
            canonical_hashes: Set[str] = set()
            for block in self.get_canonical_chain():
                if block.height >= min_height:
                    canonical_hashes.add(block.hash)

            # Find blocks to prune
            to_prune = [
                hash for hash, block in self._blocks.items()
                if block.height < min_height and hash not in canonical_hashes
            ]

            # Prune
            for hash in to_prune:
                del self._blocks[hash]
                if hash in self._scores:
                    del self._scores[hash]
                if hash in self._children:
                    del self._children[hash]

                # Remove from parent's children list
                for parent_children in self._children.values():
                    if hash in parent_children:
                        parent_children.remove(hash)

            return len(to_prune)

    def get_fork_blocks(self) -> List[BlockInfo]:
        """
        Get all blocks that are NOT on the canonical chain (fork blocks).

        Returns:
            List of non-canonical blocks
        """
        with self._lock:
            canonical_hashes = {b.hash for b in self.get_canonical_chain()}
            return [
                block for block in self._blocks.values()
                if block.hash not in canonical_hashes
            ]

    def get_common_ancestor(self, hash1: str, hash2: str) -> Optional[BlockInfo]:
        """
        Find the common ancestor of two blocks.

        Returns:
            Common ancestor block or None if not found
        """
        with self._lock:
            # Build ancestor set for hash1
            ancestors1: Set[str] = set()
            current = hash1
            while current in self._blocks:
                ancestors1.add(current)
                block = self._blocks[current]
                current = block.parent_hash

            # Walk back from hash2 until we find a common ancestor
            current = hash2
            while current in self._blocks:
                if current in ancestors1:
                    return self._blocks[current]
                block = self._blocks[current]
                current = block.parent_hash

            return None


# Module exports
__all__ = [
    "BlockInfo",
    "ForkChoice",
    "ForkChoiceError",
]

File: luxtensor/crates/luxtensor-consensus/src/fork_choice.rs
use crate::error::ConsensusError;
use luxtensor_core::block::Block;
use luxtensor_core::types::Hash;
use std::collections::{HashMap, HashSet, VecDeque};
use parking_lot::RwLock;

/// Fork choice rule implementation (GHOST algorithm)
/// Greedy Heaviest-Observed Sub-Tree
pub struct ForkChoice {
    /// All known blocks indexed by hash
    blocks: RwLock<HashMap<Hash, Block>>,
    /// Block scores (higher is better)
    scores: RwLock<HashMap<Hash, u64>>,
    /// Current head of the chain
    head: RwLock<Hash>,
    /// Genesis block hash
    genesis_hash: Hash,
}

impl ForkChoice {
    /// Create a new fork choice instance with genesis block
    pub fn new(genesis: Block) -> Self {
        let genesis_hash = genesis.hash();
        let mut blocks = HashMap::new();
        let mut scores = HashMap::new();

        blocks.insert(genesis_hash, genesis);
        scores.insert(genesis_hash, 0);

        Self {
            blocks: RwLock::new(blocks),
            scores: RwLock::new(scores),
            head: RwLock::new(genesis_hash),
            genesis_hash,
        }
    }

    /// Add a new block to the fork choice
    pub fn add_block(&self, block: Block) -> Result<(), ConsensusError> {
        let block_hash = block.hash();
        let parent_hash = block.header().previous_hash;

        // Check if we already have this block
        {
            let blocks = self.blocks.read();
            if blocks.contains_key(&block_hash) {
                return Err(ConsensusError::DuplicateBlock(block_hash));
            }
        }

        // Verify parent exists
        {
            let blocks = self.blocks.read();
            if !blocks.contains_key(&parent_hash) && parent_hash != [0u8; 32] {
                return Err(ConsensusError::OrphanBlock {
                    block: block_hash,
                    parent: parent_hash,
                });
            }
        }

        // Calculate score for new block (parent score + 1)
        let score = {
            let scores = self.scores.read();
            if parent_hash == [0u8; 32] {
                1 // Genesis block
            } else {
                scores.get(&parent_hash).copied().unwrap_or(0) + 1
            }
        };

        // Add block and score
        {
            let mut blocks = self.blocks.write();
            let mut scores = self.scores.write();
            blocks.insert(block_hash, block);
            scores.insert(block_hash, score);
        }

        // Update head if necessary
        self.update_head();

        Ok(())
    }

    /// Get the current head block
    pub fn get_head(&self) -> Result<Block, ConsensusError> {
        let head_hash = *self.head.read();
        let blocks = self.blocks.read();

        blocks
            .get(&head_hash)
            .cloned()
            .ok_or(ConsensusError::BlockNotFound(head_hash))
    }

    /// Get the head hash
    pub fn get_head_hash(&self) -> Hash {
        *self.head.read()
    }

    /// Get a block by hash
    pub fn get_block(&self, hash: &Hash) -> Result<Block, ConsensusError> {
        let blocks = self.blocks.read();
        blocks
            .get(hash)
            .cloned()
            .ok_or(ConsensusError::BlockNotFound(*hash))
    }

    /// Get the canonical chain from genesis to head
    pub fn get_canonical_chain(&self) -> Vec<Block> {
        let mut chain = Vec::new();
        let blocks = self.blocks.read();
        let mut current_hash = *self.head.read();

        // Walk backwards from head to genesis
        while current_hash != [0u8; 32] {
            if let Some(block) = blocks.get(&current_hash) {
                chain.push(block.clone());
                current_hash = block.header().previous_hash;
            } else {
                break;
            }
        }

        // Reverse to get genesis -> head order
        chain.reverse();
        chain
    }

    /// Get the score of a block
    pub fn get_score(&self, hash: &Hash) -> Option<u64> {
        let scores = self.scores.read();
        scores.get(hash).copied()
    }

    /// Check if a block exists
    pub fn has_block(&self, hash: &Hash) -> bool {
        let blocks = self.blocks.read();
        blocks.contains_key(hash)
    }

    /// Get the number of blocks in the fork choice
    pub fn block_count(&self) -> usize {
        let blocks = self.blocks.read();
        blocks.len()
    }

    /// Update the head to the block with the highest score (GHOST algorithm)
    fn update_head(&self) {
        let scores = self.scores.read();

        // Find the block with the highest score
        if let Some((&new_head, _)) = scores.iter().max_by_key(|(_, &score)| score) {
            let mut head = self.head.write();
            *head = new_head;
        }
    }

    /// Recompute scores for all blocks (useful after reorganization)
    /// Uses BFS from genesis for correct topological ordering
    pub fn recompute_scores(&self) {
        let blocks = self.blocks.read();
        let mut scores = self.scores.write();

        // Clear existing scores
        scores.clear();
        scores.insert(self.genesis_hash, 0);

        // Build parent -> children index for BFS
        let mut children: HashMap<Hash, Vec<Hash>> = HashMap::new();
        for (hash, block) in blocks.iter() {
            let parent_hash = block.header().previous_hash;
            children.entry(parent_hash).or_default().push(*hash);
        }

        // BFS from genesis ensures parents are scored before children
        let mut queue = VecDeque::new();
        queue.push_back(self.genesis_hash);

        while let Some(current) = queue.pop_front() {
            let current_score = scores.get(&current).copied().unwrap_or(0);

            if let Some(child_hashes) = children.get(&current) {
                for child_hash in child_hashes {
                    scores.insert(*child_hash, current_score + 1);
                    queue.push_back(*child_hash);
                }
            }
        }

        drop(scores);
        self.update_head();
    }

    /// Get all blocks at a specific height
    pub fn get_blocks_at_height(&self, height: u64) -> Vec<Block> {
        let blocks = self.blocks.read();
        blocks
            .values()
            .filter(|block| block.height() == height)
            .cloned()
            .collect()
    }

    /// Prune old blocks (keep only canonical chain + recent forks)
    /// Optimized: Uses HashSet for O(1) lookups instead of O(n) Vec::contains
    pub fn prune(&self, keep_depth: u64) -> Result<usize, ConsensusError> {
        let head = self.get_head()?;
        let head_height = head.height();

        if head_height < keep_depth {
            return Ok(0);
        }

        let prune_before_height = head_height - keep_depth;
        let canonical_chain = self.get_canonical_chain();

        // Use HashSet for O(1) lookup (was O(n) with Vec::contains)
        let mut keep_set: HashSet<Hash> = canonical_chain
            .iter()
            .filter(|block| block.height() >= prune_before_height)
            .map(|block| block.hash())
            .collect();

        // Also keep recent fork blocks
        {
            let blocks = self.blocks.read();
            for block in blocks.values() {
                if block.height() >= prune_before_height {
                    keep_set.insert(block.hash());
                }
            }
        }

        // Remove blocks not in keep set - O(n) instead of O(n²)
        let mut blocks = self.blocks.write();
        let mut scores = self.scores.write();
        let initial_count = blocks.len();

        blocks.retain(|hash, _| keep_set.contains(hash));
        scores.retain(|hash, _| keep_set.contains(hash));

        let pruned_count = initial_count - blocks.len();
        Ok(pruned_count)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use luxtensor_core::block::{Block, BlockHeader};

    fn create_test_block(height: u64, previous_hash: Hash) -> Block {
        let header = BlockHeader::new(
            1,                           // version
            height,
            1000 + height,               // timestamp
            previous_hash,
            [0u8; 32],                   // state_root
            [0u8; 32],                   // txs_root
            [0u8; 32],                   // receipts_root
            [0u8; 32],                   // validator (32 bytes)
            [0u8; 64],                   // signature
            0,                           // gas_used
            1000000,                     // gas_limit
            vec![],                      // extra_data
        );
        Block::new(header, vec![])
    }

    #[test]
    fn test_fork_choice_creation() {
        let genesis = create_test_block(0, [0u8; 32]);
        let fork_choice = ForkChoice::new(genesis);

        assert_eq!(fork_choice.block_count(), 1);
        assert!(fork_choice.get_head().is_ok());
    }

    #[test]
    fn test_add_block() {
        let genesis = create_test_block(0, [0u8; 32]);
        let genesis_hash = genesis.hash();
        let fork_choice = ForkChoice::new(genesis);

        let block1 = create_test_block(1, genesis_hash);
        assert!(fork_choice.add_block(block1).is_ok());
        assert_eq!(fork_choice.block_count(), 2);
    }

    #[test]
    fn test_add_duplicate_block() {
        let genesis = create_test_block(0, [0u8; 32]);
        let genesis_hash = genesis.hash();
        let fork_choice = ForkChoice::new(genesis);

        let block1 = create_test_block(1, genesis_hash);
        let block1_copy = block1.clone();

        assert!(fork_choice.add_block(block1).is_ok());
        assert!(fork_choice.add_block(block1_copy).is_err());
    }

    #[test]
    fn test_add_orphan_block() {
        let genesis = create_test_block(0, [0u8; 32]);
        let fork_choice = ForkChoice::new(genesis);

        // Block with non-existent parent
        let orphan = create_test_block(1, [1u8; 32]);
        assert!(fork_choice.add_block(orphan).is_err());
    }

    #[test]
    fn test_get_canonical_chain() {
        let genesis = create_test_block(0, [0u8; 32]);
        let genesis_hash = genesis.hash();
        let fork_choice = ForkChoice::new(genesis);

        // Add 3 blocks in sequence
        let block1 = create_test_block(1, genesis_hash);
        let block1_hash = block1.hash();
        fork_choice.add_block(block1).unwrap();

        let block2 = create_test_block(2, block1_hash);
        let block2_hash = block2.hash();
        fork_choice.add_block(block2).unwrap();

        let block3 = create_test_block(3, block2_hash);
        fork_choice.add_block(block3).unwrap();

        let chain = fork_choice.get_canonical_chain();
        assert_eq!(chain.len(), 4); // genesis + 3 blocks
        assert_eq!(chain[0].height(), 0);
        assert_eq!(chain[3].height(), 3);
    }

    #[test]
    fn test_fork_selection() {
        let genesis = create_test_block(0, [0u8; 32]);
        let genesis_hash = genesis.hash();
        let fork_choice = ForkChoice::new(genesis);

        // Create main chain: genesis -> block1 -> block2
        let block1 = create_test_block(1, genesis_hash);
        let block1_hash = block1.hash();
        fork_choice.add_block(block1).unwrap();

        let block2 = create_test_block(2, block1_hash);
        fork_choice.add_block(block2).unwrap();

        // Create fork: genesis -> block1_alt -> block2_alt -> block3_alt
        let mut block1_alt = create_test_block(1, genesis_hash);
        // Modify to make it different
        block1_alt.header_mut().timestamp += 1;
        let block1_alt_hash = block1_alt.hash();
        fork_choice.add_block(block1_alt).unwrap();

        let block2_alt = create_test_block(2, block1_alt_hash);
        let block2_alt_hash = block2_alt.hash();
        fork_choice.add_block(block2_alt).unwrap();

        let block3_alt = create_test_block(3, block2_alt_hash);
        fork_choice.add_block(block3_alt).unwrap();

        // The longer chain (3 blocks) should be selected
        let head = fork_choice.get_head().unwrap();
        assert_eq!(head.height(), 3);
    }

    #[test]
    fn test_get_blocks_at_height() {
        let genesis = create_test_block(0, [0u8; 32]);
        let genesis_hash = genesis.hash();
        let fork_choice = ForkChoice::new(genesis);

        // Add two blocks at height 1 (fork)
        let block1a = create_test_block(1, genesis_hash);
        fork_choice.add_block(block1a).unwrap();

        let mut block1b = create_test_block(1, genesis_hash);
        block1b.header_mut().timestamp += 1;
        fork_choice.add_block(block1b).unwrap();

        let blocks_at_height_1 = fork_choice.get_blocks_at_height(1);
        assert_eq!(blocks_at_height_1.len(), 2);
    }

    #[test]
    fn test_has_block() {
        let genesis = create_test_block(0, [0u8; 32]);
        let genesis_hash = genesis.hash();
        let fork_choice = ForkChoice::new(genesis);

        assert!(fork_choice.has_block(&genesis_hash));
        assert!(!fork_choice.has_block(&[1u8; 32]));
    }
}
